src/state.py

```diff
@@ -0,0 +1,128 @@
+# Copyright 2024 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""The charm state and configurations."""
+
+import enum
+import json
+import logging
+import typing
+
+import ops
+import pydantic
+
+from errors import ConfigurationError, IntegrationDataError
+
+logger = logging.getLogger(__name__)
+
+CACHE_CONFIG_INTEGRATION_NAME = "cache-config"
+
+LOCATION_CONFIG_NAME = "location"
+BACKENDS_CONFIG_NAME = "backends"
+PROTOCOL_CONFIG_NAME = "protocol"
+
+
+class Protocol(str, enum.Enum):
+    """Protocol to request backends.
+
+    Attributes:
+        HTTP: Use HTTP for requests.
+        HTTPS: Use HTTPS for requests.
+    """
+
+    HTTP = "http"
+    HTTPS = "https"
+
+
+class LocationConfig(pydantic.BaseModel):
+    """Represents the configuration for a location.
+
+    Attributes:
+        location: Defines what URL to match for this set of configuration.
+        backends: The backends for this set of configuration.
+        protocol: The protocol to request the backends with. Can be http or
+            https.
+    """
+
+    location: typing.Annotated[str, pydantic.StringConstraints(min_length=1)]
+    backends: tuple[pydantic.IPvAnyAddress, ...]
+    protocol: Protocol
+
+    @classmethod
+    def from_integration_data(cls, data: ops.RelationDataContent) -> "LocationConfig":
+        """Initialize object from the charm.
+
+        Args:
+            data: One set of integration data.
+
+        Raises:
+            ConfigurationError: Invalid cache configurations in integration data.
+
+        Returns:
+            The object.
+        """
+        location = data.get(LOCATION_CONFIG_NAME, "").strip()
```

Comment:
Beware that this location is used in the nginx `location` attribute, that only supports path. We should parse it